ydb/core/grpc_services/rpc_read_rows.cpp

```diff
@@ -148,51 +148,35 @@ class TReadRowsRPC : public TActorBootstrapped<TReadRowsRPC> {
             }
             const auto& colInfo = *colInfoPtr;
 
-            i32 typmod = -1;
-            if (typeInProto.type_id()) {
-                // TODO check Arrow types
-            } else if (typeInProto.has_decimal_type() && colInfo.PType.GetTypeId() == NScheme::NTypeIds::Decimal) {
-                int precision = typeInProto.decimal_type().precision();
-                int scale = typeInProto.decimal_type().scale();
-                if (precision != NScheme::DECIMAL_PRECISION || scale != NScheme::DECIMAL_SCALE) {
-                    errorMessage = Sprintf("Unsupported Decimal(%d,%d) for column %s: expected Decimal(%d,%d)",
-                                           precision, scale,
-                                           name.c_str(),
-                                           NScheme::DECIMAL_PRECISION, NScheme::DECIMAL_SCALE);
+            TString columnTypeName = NScheme::TypeName(colInfo.PType, colInfo.PTypeMod);
 
-                    return false;
-                }
-            } else if (typeInProto.has_pg_type()) {
-                const auto& typeName = typeInProto.pg_type().type_name();
-                auto typeDesc = NPg::TypeDescFromPgTypeName(typeName);
-                if (!typeDesc) {
-                    errorMessage = Sprintf("Unknown pg type for column %s: %s",
-                                           name.c_str(), typeName.c_str());
-                    return false;
-                }
+            TString parseProtoError;
+            NScheme::TTypeInfo typeInfo;
+            if (!NScheme::TypeInfoFromProto(typeInProto, typeInfo, parseProtoError)){
+                errorMessage = Sprintf("Type parse error for column %s: %s",
+                    name.c_str(), parseProtoError.c_str());
+                return false;
+            }
 
-                const auto typeInRequest = NScheme::TTypeInfo(typeDesc);
-                if (typeInRequest != colInfo.PType) {
-                    errorMessage = Sprintf("Type mismatch for column %s: expected %s, got %s",
-                                           name.c_str(), NScheme::TypeName(colInfo.PType).c_str(),
-                                           NScheme::TypeName(typeInRequest).c_str());
-                    return false;
-                }
+            TString inTypeName = NScheme::TypeName(typeInfo, typeInfo.GetPgTypeMod(colInfo.PTypeMod));
+
+            if (typeInfo != colInfo.PType) {
+                errorMessage = Sprintf("Unexpected type %s for column %s: expected %s",
```

Comment:
NIT: For me "Type mismatch for column" from old code is better message

Reply:
0b2b82fab577a42e687f3e1d0b7ea3dbd5e26c32